modules/indexing/src/main/java/org/apache/ignite/internal/processors/query/h2/sql/SplitterUtils.java

```diff
@@ -286,6 +290,255 @@ public static boolean hasPartitionedTables(GridSqlAst ast) {
         return false;
     }
 
+    /**
+     * Traverse AST while join operation isn't found. Check it if found.
+     */
+    public static void lookForPartitionedJoin(GridSqlAst ast, GridSqlAst upWhere, IgniteLogger log) {
```

Comment:
I'm really struggling to get a sense of completeness from this code. The test cases list seems extensive, so this code probably works, but I'm not sure I understand why...

I think structuring might solve it.

How about the following structure?

1. We always start from a SELECT - so let's process it first
2. In a SELECT, we want to check the subselects in the columns (yes!), the FROM, and the subselects in the where
3. In a FROM, we may see either a single table (not interesting), a subselect (go to step 1), or a JOIN

... and so on. I don't insist on a particular structure but I think we need to be able to understand from the code that nothing is missing. BTW splitting into several methods (lookForPartitionJoinInSelect/From/Where/etc) could help.

Reply:
I added more comments in this method. Now it should be more straightforward. WDYT?